### PR TITLE
*: allow dynamic link openssl library

### LINKED ISSUE #15943: Build TiKV that dynamically link OpenSSL library
## Development Task

Currently, TiKV binaries are statically linked to the OpenSSL library, preventing the use of the host system's OpenSSL.

We need to build TiKV with dynamic linking of the OpenSSL library. This is particularly useful in FIPS scenarios where TiKV needs to delegate cryptographic operations to the host FIPS OpenSSL.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #15943

Requires https://github.com/tikv/rust-rocksdb/pull/774

What's Changed:

```commit-message
Currently, TiKV binaries are statically linked to the OpenSSL library,
preventing the use of the host system's OpenSSL.

This commit adds an option to build TiKV with dynamic linking of the
OpenSSL library, enabling TiKV to utilize the host system's OpenSSL.
This is particularly useful in FIPS scenarios where TiKV needs to
delegate cryptographic operations to the host FIPS OpenSSL.

By default, this option is disabled, and TiKV continues to statically
link the OpenSSL library.
```

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

`ENABLE_FIPS=0 make docker`

### Release note 

```release-note
none
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -379,10 +390,9 @@ opt-level = 1
 [profile.dev.package.libtitan_sys]
 debug = false
 opt-level = 1
-
-[profile.dev.package.tirocks-sys]
-debug = false
-opt-level = 1
+# [profile.dev.package.tirocks-sys]
```
Comment (Connor1996) [inline Cargo.toml:393]: clean?

File: Cargo.toml [hunk 2]
```diff
@@ -403,7 +413,7 @@ rpath = false
 opt-level = 3
 debug = false
 codegen-units = 1
-lto = "thin"
+lto = false
```
Comment (Connor1996) [inline Cargo.toml:416]: why

File: Cargo.toml [hunk 3]
```diff
@@ -31,6 +31,17 @@ test-engine-raft-raft-engine = ["engine_test/test-engine-raft-raft-engine"]
 test-engines-rocksdb = ["engine_test/test-engines-rocksdb"]
 test-engines-panic = ["engine_test/test-engines-panic"]
 pprof-fp = ["pprof/frame-pointer"]
+openssl-vendored = [
```
Comment (Connor1996) [inline Cargo.toml:34]: missing `etcd-client/tls-openssl-vendored`?

Comment (overvenus) [reply Cargo.toml:34]: I find `etcd-client` is used some dead code in `backup-stream` crate. These dead code and the `etcd-client` are guarded by `metastore-etcd` features and will not be compiled by default.

Comment (Connor1996) [reply Cargo.toml:34]: okay, please add a comment

**Review by tonyxuqqi — APPROVED** [review]

File: Cargo.toml (hunk 3 above)
Comment (overvenus) [reply Cargo.toml:34]: Removed by https://github.com/tikv/tikv/pull/15976

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**